src/shared/Microsoft.Git.CredentialManager.Tests/HostProviderRegistryTests.cs

```diff
@@ -57,5 +83,140 @@ public void HostProviderRegistry_MultipleValidProviders_ReturnsFirstRegistered()
+            provider3Mock.Setup(x => x.Id).Returns("provider3");
+            provider3Mock.Setup(x => x.IsSupported(It.IsAny<InputArguments>())).Returns(false);
+
+            registry.Register(provider1Mock.Object, provider2Mock.Object, provider3Mock.Object);
+
+            IHostProvider result = registry.GetProvider(input);
+
+            Assert.Same(provider3Mock.Object, result);
+        }
+
+        [Fact]
+        public void HostProviderRegistry_GetProvider_AutoProviderSpecified_ReturnsFirstSupportedProvider()
+        {
+            var context = new TestCommandContext
+            {
+                Settings = {ProviderOverride = Constants.ProviderIdAuto}
+            };
+            var registry = new HostProviderRegistry(context);
+            var input = new InputArguments(new Dictionary<string, string>());
+
+            var provider1Mock = new Mock<IHostProvider>();
+            var provider2Mock = new Mock<IHostProvider>();
+            var provider3Mock = new Mock<IHostProvider>();
+            provider1Mock.Setup(x => x.Id).Returns("provider1");
+            provider1Mock.Setup(x => x.IsSupported(It.IsAny<InputArguments>())).Returns(false);
+            provider2Mock.Setup(x => x.Id).Returns("provider2");
+            provider2Mock.Setup(x => x.IsSupported(It.IsAny<InputArguments>())).Returns(true);
+            provider3Mock.Setup(x => x.Id).Returns("provider3");
+            provider3Mock.Setup(x => x.IsSupported(It.IsAny<InputArguments>())).Returns(false);
+
+            registry.Register(provider1Mock.Object, provider2Mock.Object, provider3Mock.Object);
+
+            IHostProvider result = registry.GetProvider(input);
+
+            Assert.Same(provider2Mock.Object, result);
+        }
+
+        [Fact]
+        public void HostProviderRegistry_GetProvider_UnknownProviderSpecified_ReturnsFirstSupportedProvider()
+        {
+            var context = new TestCommandContext
+            {
+                Settings = {ProviderOverride = "provider42"}
+            };
+            var registry = new HostProviderRegistry(context);
+            var input = new InputArguments(new Dictionary<string, string>());
+
+            var provider1Mock = new Mock<IHostProvider>();
+            var provider2Mock = new Mock<IHostProvider>();
+            var provider3Mock = new Mock<IHostProvider>();
+            provider1Mock.Setup(x => x.Id).Returns("provider1");
+            provider1Mock.Setup(x => x.IsSupported(It.IsAny<InputArguments>())).Returns(false);
+            provider2Mock.Setup(x => x.Id).Returns("provider2");
+            provider2Mock.Setup(x => x.IsSupported(It.IsAny<InputArguments>())).Returns(true);
+            provider3Mock.Setup(x => x.Id).Returns("provider3");
+            provider3Mock.Setup(x => x.IsSupported(It.IsAny<InputArguments>())).Returns(false);
+
+            registry.Register(provider1Mock.Object, provider2Mock.Object, provider3Mock.Object);
+
+            IHostProvider result = registry.GetProvider(input);
```

Comment:
So if they specify a provider that doesn't exist, we default to the first true?  Would there be a warning?

Reply:
We print the following to the trace output:

```
No host provider was found with ID '{providerId}'.. falling back to auto-detection.;
```

Perhaps we should also write to stderr...

Reply:
Added stderr warning message.